content/blog/posts/o3de-graphics-atom-intro.md

```diff
@@ -0,0 +1,35 @@
+
+I began by simply editing a vertex shader of the *StandardPBR* material type to display the object at an offset from the original position. All I added was `position.x += 5`.
+
+{{< image-width src="/images/blog/graphics-atom-intro/offset.png" width="100%" alt="The step of drawing the shaderball at an offset" >}}
+
+I was ecstatic to see something show up at an offset from just this one line of shader code. However, I was quite confused as to why there was a gray outline of the material at the original position, with the texture offset as expected. This was when I realized the importance of passes. StandardPBR has many passes, with each pass rendering a different part of the material. In my case above, I had edited the vertex shader for the forward pass, while the depth pass still used an unedited vertex shader. These sequential passes were necessary because the outputs of one could be used as an input for another.
+
+From there, I started writing my own material type. I referenced the shader code in a .azsl file via entry points in a .shader file, which itself was referenced in the overarching .materialtype file. This gave me a broad view of how Atom renders each material and also taught me the function of each shader resource group, function, pass, and more. While the number of files and connections were intimidating at first, once I became familiar with the terms and what each component did, I appreciated Atom’s pipeline organization and code modularization.
+
+An interesting aspect of graphics that I noticed is the focus on optimization. While many fields of computer science promote optimization, graphics optimization appears particularly important because a low frame rate is obvious. Also, since graphics rely heavily on vector math and physics, optimizing graphics code isn’t necessarily limited to choosing which data structure or algorithm to use, but also which mathematical model. A simple example that I ran into was when I needed to use an oscillation function to mimic vegetation bending. The most obvious function is the sine wave. However, I learned that using sine is considered relatively slow, so it is more ideal to use a sine lookup table or to use triangle waves.
+
+## Looking back
+
+{{< image-width src="/images/blog/graphics-atom-intro/tree.png" width="100%" alt="The progress of making a tree." >}}
+
+As I added more shader code and more passes, I saw my material evolving to eventually actually look like a tree and then sway! That, to me, was one of the most rewarding parts of using O3DE and writing shader code: the almost immediate visual feedback. Additionally, looking back at the process, I appreciate the beauty of how everything fits together. My material type uses multiple passes, with each pass editing a particular part of the material via my custom shaders.
```

Comment:
Suggested rewrite: As I added more shader code and more passes, I saw my material evolving, eventually looking like an actual tree and then swaying!